### PR TITLE
dbt-materialize: replace macros with custom materializations

### PR DESCRIPTION
**Pull request by JLDLaughlin**
Background and motivation: https://github.com/MaterializeInc/materialize/issues/7810.

This PR removes all of the `mz_create_object` and `mz_drop_object` macros and replaces them with three new types of custom dbt [materializations](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/materializations): `source`, `index`, and `sink`. This solves the problem described in #7810 because `dbt docs generate` executes the macros in each of the model files, but does not execute the models themselves. By removing the macros, sources, indexes, and sinks, will not be dropped arbitrarily by the `dbt docs` commands. As a bonus, this change will also make our dbt-materialize implementation more in line with dbt's [model](https://docs.getdbt.com/docs/building-a-dbt-project/building-models) ethos.

Note: This fixes the open issues listed in #7810. However, it doesn't yet include column information for indexes. I'll do that in a follow up PR.

### CODE REVIEW

**Comment by JLDLaughlin** [comment]
@mjibson & @sploiselle -- please let me know if anything in the connected issue or this PR is unclear! happy to chat through this more on Github or on Zoom.

File: misc/dbt-materialize/dbt/adapters/materialize/relation.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional
+
+from dbt.adapters.postgres import PostgresRelation
+from mashumaro.types import SerializableType
+
+
+# Stolen from https://github.com/dbt-labs/dbt/blob/develop/core/dbt/dataclass_schema.py,
```
Comment (JLDLaughlin) [inline misc/dbt-materialize/dbt/adapters/materialize/relation.py:24]: @quodlibetor can you confirm that I'm not doing something silly?

Comment (quodlibetor) [reply misc/dbt-materialize/dbt/adapters/materialize/relation.py:24]: _micro_ nit: the comma in this is getting added to the autolinkification in the github ui and making it look like a 404, maybe add a space?

File: misc/dbt-materialize/dbt/adapters/materialize/relation.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# which is not importable.
+class StrEnum(str, SerializableType, Enum):
+    def __str__(self):
+        return self.value
+
+    # https://docs.python.org/3.6/library/enum.html#using-automatic-values
+    def _generate_next_value_(name, *_):
+        return name
```
Comment (quodlibetor) [inline misc/dbt-materialize/dbt/adapters/materialize/relation.py:32]: This is fine, but seems to be ignored. If we changed this to `return name.lower()` then this below:

```python
class MaterializeRelationType(StrEnum):
    Source = "source"
    View = "view"
    MaterializedView = "materializedview"
    ...
```

should be usable as:

```python
class MaterializeRelationType(StrEnum):
    Source = auto()
    View = auto()
    MaterializedView = auto()
    ...
```

not sure if that's better, though.

**Review by quodlibetor — COMMENTED** [review]
I've only looked at the python code, but it LGTM! Small comments, but neither of them need to be addressed.

**Review by quodlibetor — APPROVED** [review]
I believe based on this PR (specifically the examples) that the `macros/materializations/*.sql` files intercept the `CREATE <foo> ...` calls when used in models.

If so, I bet that that is obvious based on the docs that you link to in the PR description to folks that are familiar with dbt, but maybe it could be called out explicitly in the PR description/commit message? Specifically, it seems like dbt executes all macros but not all models for docs? If I'm wrong about that then my code review is probably not valid and I should have this explained to me.

**Comment by JLDLaughlin** [comment]
So the `materialization` bit is what determines which `macros/materializations/*.sql` file is called (`materializedview` goes to `materialized_view.sql`, etc.). You're right that all macros are executed during `dbt docs generate`, but not the models. I can update the commit message and the PR description!

**Comment by JLDLaughlin** [comment]
@sploiselle & @mjibson -- I'm planning on merging and releasing this change tomorrow unless you all have objections! also happy to talk through anything in more detail, either via comments here or Slack/Zoom.

**Comment by sploiselle** [comment]
@JLDLaughlin Sorry for the radio silence here! I have time slated tomorrow to learn more about DBT, but as of right now I am pretty clueless as to what it's aiming to do. If you have time to burn tomorrow, would be glad for a walkthrough, but also very happy to beat on it myself.

**Pull request merged**